File: src/java/me/indexyz/strap/utils/Configuration.java
package me.indexyz.strap.utils;

import java.io.*;
import java.util.Date;
import java.util.Properties;

public class Configuration {
    private static final String CONFIG_NAME = "config.properties";
    private Properties configs;

    private Configuration() {
        this.reload();
    }
    private static Configuration instance = null;

    public static Configuration get() {
        if (Configuration.instance == null) {
            Configuration.instance = new Configuration();
        }

        return Configuration.instance;
    }

    private void reload() {
        try {
            FileReader reader = new FileReader(CONFIG_NAME);
            Properties properties = new Properties();

            properties.load(reader);
            configs = properties;
        } catch (IOException e) {
            this.createDefault();
        }
    }

    public void save() {
        try {
            FileWriter writer = new FileWriter(CONFIG_NAME);
            Date date = new Date();
            this.configs.store(writer, "Strap bot config file");
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private void createDefault() {
        var properties = new Properties();
        properties.setProperty("botToken", "");
        configs = properties;

        this.save();
    }

    public String get(String key) {
        var envs = System.getProperties();
        if (envs.containsKey(key.toUpperCase())) {
            return envs.getProperty(key.toUpperCase());
        }
        return configs.getProperty(key);
    }

    public boolean has(String key) {
        var envs = System.getProperties();
        if (envs.containsKey(key.toUpperCase())) {
            return true;
        }
        return configs.containsKey(key);
    }

    public void set(String key, String value) {
        configs.setProperty(key, value);
    }
}

File: src/kotlin/me/indexyz/strap/utils/Configuration.kt
package me.indexyz.strap.utils

import java.io.FileReader
import java.io.FileWriter
import java.io.IOException
import java.util.*

class Configuration() {
    private val CONFIG_NAME = "config.properties"
    private var configs: Properties? = null

    init {
        this.reload()
    }

    companion object {
        private var instance: Configuration? = null
            get() {
                if (field == null) {
                    field = Configuration()
                }
                return field
            }
        fun get(): Configuration {
            return instance!!
        }
    }


    private fun reload() {
        try {
            val reader = FileReader(CONFIG_NAME)
            val properties = Properties()

            properties.load(reader)
            configs = properties
        } catch (e: IOException) {
            this.createDefault()
        }

    }

    fun save() {
        try {
            val writer = FileWriter(CONFIG_NAME)
            val date = Date()
            this.configs!!.store(writer, "Strap bot config file")
        } catch (e: IOException) {
            e.printStackTrace()
        }

    }

    private fun createDefault() {
        val properties = Properties()
        properties.setProperty("botToken", "")
        configs = properties

        this.save()
    }

    operator fun get(key: String): String {
        val envs = System.getProperties()
        return if (envs.containsKey(key.toUpperCase())) {
            envs.getProperty(key.toUpperCase())
        } else configs!!.getProperty(key)
    }

    fun has(key: String): Boolean {
        val envs = System.getProperties()
        return if (envs.containsKey(key.toUpperCase())) {
            true
        } else configs!!.containsKey(key)
    }

    operator fun set(key: String, value: String) {
        configs!!.setProperty(key, value)
    }
}